### PR TITLE
Purging state on leave, with exceptions. Applied shfmt to the script.

### PR DESCRIPTION
**Pull request by isavcic**
Signed-off-by: Ivan Savcic <isavcic@gmail.com>

### CODE REVIEW

**Comment by isavcic** [comment]
Rationale for Kubernetes: on StatefulSet scale down, persistent volumes remain, yet the pods have left the cluster. On subsequent scale up, those volumes are reused, and if there is state on them, they won't be able to join the cluster. For those purposes, we want to purge the state on leave, and double check on startup whether the state is clean when the pod ID is unknown to the cluster (ie. it had previously left). We are also checking if we're the last pod in the cluster, since we won't be able to migrate the data off in that case, so in that case we're just stopping VerneMQ and leaving the state as-is.

**Comment by ioolkos** [comment]
@isavcic thanks for the PR and the explanation! 

I think this would be a clean solution for the 'leaving' case (highly improved on the existing script!), supposed the leaving node manages to replicate it's offline queues to the remaining cluster nodes in time.

If I'm correct, it does not resolve Pod re-scheduling, and this is currently the main issue in production: Kubernetes just terminates Verne nodes, and for those cases it would be cleaner and faster to just stop them, let them keep their state and then just restart them.

**Comment by isavcic** [comment]
@ioolkos I believe I've also tackled that now with:

```bash
statefulset=$(curl -X GET --cacert /var/run/secrets/kubernetes.io/serviceaccount/ca.crt -H "Authorization: Bearer $(cat /var/run/secrets/kubernetes.io/serviceaccount/token)" \
    https://kubernetes.default.svc.$DOCKER_VERNEMQ_KUBERNETES_CLUSTER_NAME/api/v1/namespaces/$(cat /var/run/secrets/kubernetes.io/serviceaccount/namespace)/pods/$(hostname) | jq -r '.metadata.ownerReferences[0].name')

reschedule=$(curl -X GET --cacert /var/run/secrets/kubernetes.io/serviceaccount/ca.crt -H "Authorization: Bearer $(cat /var/run/secrets/kubernetes.io/serviceaccount/token)" \
    https://kubernetes.default.svc.$DOCKER_VERNEMQ_KUBERNETES_CLUSTER_NAME/apis/apps/v1/namespaces/$(cat /var/run/secrets/kubernetes.io/serviceaccount/namespace)/statefulsets/${statefulset} | jq '.status.replicas == .status.currentReplicas')

if [[ $reschedule == "true" ]]; then
    echo "Reschedule is true, not leaving the cluster"
    /vernemq/bin/vmq-admin node stop >/dev/null
else
    echo "Reschedule is false, leaving the cluster"
    /vernemq/bin/vmq-admin cluster leave node=$terminating_node_name -k && rm -rf /vernemq/data/*
fi
```

**Comment by ioolkos** [comment]
@isavcic brilliant :) wonder whether the API calls could race or timeout. I'll try to test things a little bit, and, well...at some point only production deployments can tell. Thanks!

**Comment by isavcic** [comment]
@ioolkos while I was testing this, scaling up and down and killing the pods, I've ran curl to the API in a while loop without any sleep, in other words hammered it. :) It didn't time out.

**Comment by akaltsikis** [comment]
We are also testing this branch and will report back if anything on our side is not working as expected.

**Comment by angeloskaltsikis** [comment]
We are also testing this branch and will get back with anything that will not work for us as expected. Thanks @isavcic about it 🏅

**Comment by ioolkos** [comment]
@isavcic @angeloskaltsikis one thing that we need to reflect on is what to do with Swarm or plain Docker users because they can't call K8s API.

File: bin/vernemq.sh [hunk 1]
```diff
@@ -157,8 +159,26 @@ sigterm_handler() {
...
+        if [ $kube_pod_names == $MY_POD_NAME ]; then
+            echo "I'm the only pod remaining, not performing leave and state purge."
+            /vernemq/bin/vmq-admin node stop >/dev/null
+        else
+
+            statefulset=$(curl -X GET --cacert /var/run/secrets/kubernetes.io/serviceaccount/ca.crt -H "Authorization: Bearer $(cat /var/run/secrets/kubernetes.io/serviceaccount/token)" \
+                https://kubernetes.default.svc.$DOCKER_VERNEMQ_KUBERNETES_CLUSTER_NAME/api/v1/namespaces/$(cat /var/run/secrets/kubernetes.io/serviceaccount/namespace)/pods/$(hostname) | jq -r '.metadata.ownerReferences[0].name')
+
+            reschedule=$(curl -X GET --cacert /var/run/secrets/kubernetes.io/serviceaccount/ca.crt -H "Authorization: Bearer $(cat /var/run/secrets/kubernetes.io/serviceaccount/token)" \
+                https://kubernetes.default.svc.$DOCKER_VERNEMQ_KUBERNETES_CLUSTER_NAME/apis/apps/v1/namespaces/$(cat /var/run/secrets/kubernetes.io/serviceaccount/namespace)/statefulsets/${statefulset} | jq '.status.replicas == .status.currentReplicas')
```
Comment (angeloskaltsikis) [inline bin/vernemq.sh:172]: Today we stumbled upon an issue with the current script and the helm chart it exist which is due to the service account not having the appropriate permissions to `get` `statefulsets` from `apps`.
This means we need to add the following rule
```
  - apiGroups: ["apps"]
    resources: ["statefulsets"]
    verbs: ["get"]
``` 
to the [role.yaml template](https://github.com/vernemq/docker-vernemq/blob/master/helm/vernemq/templates/role.yaml)
@isavcic can you modify the helm template to this PR just to help anyone that needs to test it.

File: bin/vernemq.sh [hunk 2]
```diff
@@ -61,19 +61,21 @@ if env | grep "DOCKER_VERNEMQ_DISCOVERY_KUBERNETES" -q; then
...
+            break
         fi
-        if [ $kube_pod_name != $MY_POD_NAME ]
-            then
-                echo "Will join an existing Kubernetes cluster with discovery node at ${kube_pod_name}.${VERNEMQ_KUBERNETES_SUBDOMAIN}.${DOCKER_VERNEMQ_KUBERNETES_NAMESPACE}.svc.${DOCKER_VERNEMQ_KUBERNETES_CLUSTER_NAME}"
-                echo "-eval \"vmq_server_cmd:node_join('VerneMQ@${kube_pod_name}.${VERNEMQ_KUBERNETES_SUBDOMAIN}.${DOCKER_VERNEMQ_KUBERNETES_NAMESPACE}.svc.${DOCKER_VERNEMQ_KUBERNETES_CLUSTER_NAME}')\"" >> /vernemq/etc/vm.args
-                break
+        if [ $kube_pod_name != $MY_POD_NAME ]; then
+            echo "Will join an existing Kubernetes cluster with discovery node at ${kube_pod_name}.${VERNEMQ_KUBERNETES_SUBDOMAIN}.${DOCKER_VERNEMQ_KUBERNETES_NAMESPACE}.svc.${DOCKER_VERNEMQ_KUBERNETES_CLUSTER_NAME}"
+            echo "-eval \"vmq_server_cmd:node_join('VerneMQ@${kube_pod_name}.${VERNEMQ_KUBERNETES_SUBDOMAIN}.${DOCKER_VERNEMQ_KUBERNETES_NAMESPACE}.svc.${DOCKER_VERNEMQ_KUBERNETES_CLUSTER_NAME}')\"" >>/vernemq/etc/vm.args
```
Comment (angeloskaltsikis) [inline bin/vernemq.sh:72]: Today during a redeployment of an already existing cluster we faced the following corner case which resulted in 2 different VerneMQ Clusters.
1st cluster
```
vernemq-0
```
2nd cluster
```
vernemq-1
vernemq-2
```

Due to `if [ $kube_pod_name != $MY_POD_NAME ]; then` the following will happen
```
vernemq-0 will try to connect to vernemq-1
vernemq-1 will try to connect to vernemq-0
vernemq-2 will try to connect to vernemq-0
```
So during the redeployment `vernemq-2` is firstly restarted and tried to connect to the node mentioned above, which is successful and continues as it should.
To continue with `vernemq-1` is then restarted and it tries to connect to the node mentioned above but faced the below
```
15:46:42.860 [info] Sent join request to: 'VerneMQ@vernemq-0.vernemq-headless.vernemq.svc.cluster.local'
15:46:42.863 [info] Unable to connect to 'VerneMQ@vernemq-0.vernemq-headless.vernemq.svc.cluster.local'
```
for some strange reason the `vernemq-0` pod has already started the update as well during this time period and faced the same error
```
15:46:50.307 [info] Sent join request to: 'VerneMQ@vernemq-1.vernemq-headless.vernemq.svc.cluster.local'
15:46:50.309 [info] Unable to connect to 'VerneMQ@vernemq-1.vernemq-headless.vernemq.svc.cluster.local'
```
For some strange reason we ended up with 2 clusters.
As soon as we deleted the `vernemq-0` it connected correctly to the correct VerneMQ cluster and joined the other nodes.
Unfortunately we didn't keep the logs of that due to being in a hurry to fix the issue.
Any thoughts?

Comment (ioolkos) [reply bin/vernemq.sh:72]: @angeloskaltsikis you're calling this re-deployment but also upgrade. What exactly did you try to do? 
Is there some sort of internal name caching or dns caching in K8s going on here maybe?

**Comment by dergraf** [comment]
@ioolkos I think this won't be an issue as the `kube_pod_names` variable gets an invalid value assigned, which will certainly not match the following if clause.

**Pull request merged**